src/OSPSuite.SimModelNative/src/Species.cpp

```diff
@@ -37,6 +37,47 @@ Species::~Species(void)
 		delete[] _RHS_UsedVariablesIndices;
 		_RHS_UsedVariablesIndices = NULL;
 	}
+	for (const auto& myPair : _jacobian_parameter_map) {
+		delete myPair.second;
+	}
+	for (const auto& myPair : _jacobian_state_variable_map) {
+		delete myPair.second;
+	}
+}
+
+Formula* Species::createJacobianFor(const int index)
+{
+	const auto& formula = DE_Jacobian(index);
+	return formula->RecursiveSimplify();
+}
+
+void Species::CacheJacobianFormulaForParameter(const int parameterIndex) 
+{
+	_jacobian_parameter_map[parameterIndex] = createJacobianFor(-parameterIndex);
+}
+
+void Species::CalculateJacobianStateVariableFor(const int stateVariableIndex)
+{
+	_jacobian_state_variable_map[stateVariableIndex] = createJacobianFor(stateVariableIndex);
+}
+
+Formula* Species::jacobianFor(const int index, const int mode, std::map<int, Formula*>* map)
+{
+	if (map->find(index) == map->end())
+	{
+		auto formula = createJacobianFor(index * mode);
+		map->insert(std::pair<int, Formula*>(index, formula));
+		return formula;
+	}
+	return map->at(index);
+}
```

Comment:
here the map is iterated **twice**: 1) `map->find(index)` 2) `map->at(index)`
I would try to iterate only once, e.g.

```
auto it = map->find(index);
if (it != map->end())
    return it->second;

auto formula = createJacobianFor(index * mode);
map->insert(std::pair<int, Formula*>(index, formula));
return formula;
```